### PR TITLE
MLIP-000

### PR DESCRIPTION
**Pull request by zurk**

### CODE REVIEW

**Comment by eiso** [comment]
Really interesting proposal and seems to organize things more logically. 
The big question seems to be, where do we train? What are your thoughts @vmarkovtsev?

File: documentation/proposals/MLIP-000.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+2. Extractors (can be inside transformers, but better to keep here)
+	1. Keep all feature extractors here
+3. Models
+	3. Keep here all asdf models
+4. Utils
+	1. Move all helpers here like bigartm, engine, token_parser, vw_dataset, etc. 
+	2. Can keep non-helpers outside: swivel, projector, …
+
+## Package content
+1. We should not have a bblfsh_meta.py with roles list and ask LA team to include it in bblfsh python package.
```
Comment (vmarkovtsev) [inline documentation/proposals/MLIP-000.md]: Agree, pls make the issue

File: documentation/proposals/MLIP-000.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
+# Source{d} ml proposal-000
+This document contains proposals and questoins to discuss for source{d} ml (https://github.com/src-d/ml).
```
Comment (vmarkovtsev) [inline documentation/proposals/MLIP-000.md]: questoins -> questions

`[source{d} ml](https://github.com/src-d/ml)`

File: documentation/proposals/MLIP-000.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+```
+self.parse_uast = importlib.import_module(
+   "bblfsh.gopkg.in.bblfsh.sdk.%s.uast.generated_pb2" % VERSION
+```
+
+### 2. Add bblfsh version check.
+As Vadim ask here https://github.com/bblfsh/client-python/issues/50. Just not to forget.
+
+### 3. What should be inside ml, and what on top of that?
+From my point of view, `src-d/ml` is something to create final datasets and provide basic transformers for all ML models.
```
Comment (vmarkovtsev) [inline documentation/proposals/MLIP-000.md]: We will discuss this, not precise

File: documentation/proposals/MLIP-000.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+For me, bag denotes zero-one features, where you don’t have any order.
+Usually, they are related to “word” information. 
+The weighted bag is the same as features set, but more or less consistent. 
+
+If we have some structural UAST features it is better not to call it bags. 
+So in `uast_struct_to_bag.py` we have feature generation, not bags. 
+
+### 5. `uast_ids_to_bag.py`, `uast_struct_to_bag.py`, `BagsExtractor`
+We have transformers to transform data and have Bags(Features)Extractor to get features. 
+It is ok and logical. But why do we have `UastIds2Bag` and `UastStructure2BagBase` with only one meaningful function: `uast_to_bag` (uast_to_features)? We don't use `vocabulary` property anymore.
```
Comment (vmarkovtsev) [inline documentation/proposals/MLIP-000.md]: We do in Apollo for example. We will discuss API levels.

File: documentation/proposals/MLIP-000.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+```
+uasts.query_uast('//*[@roleIdentifier and not(@roleIncomplete)]').extract_tokens()
+```
+2. Use `token_parser` to parse tokens
+3. Use `_vocabulary` to map tokens to something as bag keys
+4. Return bag keys frequencies.
+
+Mostly the same we have in `token_mapper.py` using the engine (In PR right now).
+
+### 7. Move `NoopTokenParser` and `HashedTokenParser` to parcers. And join with `NoTokenParser`.
```
Comment (vmarkovtsev) [inline documentation/proposals/MLIP-000.md]: I don't even remember what are they.

File: documentation/proposals/MLIP-000.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+### 8. Move `OrderedDocumentFrequencies` to models
+
+### 9. Repo2DocFreq
+Let’s include this step inside feature extraction because it just calculates docfrec independently for each feature set. Looks strange in the pipeline and not so beautiful inside.
```
Comment (vmarkovtsev) [inline documentation/proposals/MLIP-000.md]: Yep, it does only one thing and it does it very well. https://en.wikipedia.org/wiki/Single_responsibility_principle

File: documentation/proposals/MLIP-000.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+Solutions
+1. Include to `Repo2WeightedSet` Or to extractors itself
```
Comment (vmarkovtsev) [inline documentation/proposals/MLIP-000.md]: Complicates by hiding the steps

File: documentation/proposals/MLIP-000.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+2. Calculate everything independently in the pipeline and Add special `JoinFeatures` Transformer to join all features by key to one rdd. It will be easier to understand what is doing on in the pipeline.
```
Comment (vmarkovtsev) [inline documentation/proposals/MLIP-000.md]: Complicates by adding more steps

**Pull request merged**